Ac autocode dashboard

Keylas opened the pull request:
Use of [Victory](https://formidable.com/open-source/victory/) graphs to provide the teacher with some information about the activity (namely, the current success rate of the students on the tests and their use of the debug functionality)

Keylas committed:
Initial commit for ac-autocode

Keylas committed:
Split index in different file.
Unnullify dashboard.

Keylas committed:
Fix errors.

lfaucon committed:
Merge branch 'develop' into acAutogradedCode

Keylas committed:
Solve warnings issues

Keylas committed:
Merge branch 'acAutogradedCode' of https://github.com/chili-epfl/FROG into acAutogradedCode

Keylas committed:
WIP

Keylas committed:
Merge branch 'develop' into acAutogradedCode, tests for Skulpt (Python)

Keylas committed:
Sketch of a functioning ActivityRunner

Keylas committed:
Merge branch 'develop' into acAutogradedCode

Keylas committed:
Fix some issues by implementing a state for ActivityRunner

Keylas committed:
Error fixes

Keylas committed:
Merge branch 'develop' into acAutogradedCode

Keylas committed:
git fix

Keylas committed:
git fix

Keylas committed:
Merge branch 'acAutogradedCode' of https://github.com/chili-epfl/FROG into acAutogradedCode

Keylas committed:
Merge branch 'acAutogradedCode' of https://github.com/chili-epfl/FROG into acAutogradedCode

Keylas committed:
Merge branch 'acAutogradedCode' of https://github.com/chili-epfl/FROG into acAutogradedCode

Keylas committed:
Adress deepscan issue

Keylas committed:
Merge branch 'develop' into acAutogradedCode

Keylas committed:
yarn add react-ace

Keylas committed:
setup of react-ace

Keylas committed:
Merge branch 'develop' into acAutocodeFeatures

Keylas committed:
improved config, used it the rendering

Keylas committed:
exampleConfig2, placeholder rendering for tests

lfaucon committed:
Merge branch 'develop' into acAutogradedCode

Keylas committed:
eslint fix

lfaucon committed:
Use reactiveData, add template in config, and some clean up

lfaucon committed:
Merge branch 'develop' into cs411BKT

Keylas committed:
Added test support. Component breakdown.

lfaucon committed:
WIP

lfaucon committed:
WIP

lfaucon committed:
Update one example with new configuration schema

lfaucon committed:
Some layout and renaming

lfaucon committed:
Handling outputs and feedback

lfaucon committed:
Add testPanel file

lfaucon committed:
Handles errors

lfaucon committed:
Add small dashboard

lfaucon committed:
Add debug option

Keylas committed:
Update config example to fit new structure

Keylas committed:
Fix

lfaucon committed:
details

lfaucon committed:
Merge branch 'cs411BKT' of github.com:chili-epfl/FROG into cs411BKT

Keylas committed:
WIP

Keylas committed:
Use of preCode / postCode for testing, solution is run apart.

Keylas committed:
Merge branch 'cs411BKT' of https://github.com/chili-epfl/FROG into cs411BKT

Keylas committed:
Cleanup, turn rule gobal-require off

Keylas committed:
Merge branch 'develop' into cs411BKT

Keylas committed:
Merge branch 'develop' into cs411BKT

Keylas committed:
Merge branch 'cs411BKT' of https://github.com/chili-epfl/FROG into cs411BKT

Keylas committed:
WIP

Keylas committed:
WIP

Keylas committed:
Requested changes

Keylas committed:
Renaming, better handling of teacher code error

Keylas committed:
javascript runCode with proxy (WIP)

Keylas committed:
Merge branch 'cs411BKT' into acAutocodeJavascript

Keylas committed:
Javascript support, improved error handeling

Keylas committed:
Minor changes

Keylas committed:
Merge branch 'develop' into acAutocodeJavascript

Keylas committed:
Merge branch 'develop' into acAutocodeJavascript

Keylas committed:
Minor adjustments

Keylas committed:
Small modification

Keylas committed:
Logging and mergeLog

Keylas committed:
add victory (charts) to package.json

Keylas committed:
WIP

Keylas committed:
Improvements

Keylas committed:
eslint compliance

Keylas committed:
Merge branch 'develop' into acAutocodeDashboard

Keylas committed:
Minor adjustments

Keylas committed:
debug to run

lfaucon committed:
A bit of clean-up

lfaucon reviewed (commented):
@Keylas I tried to reduce the number of lines of code for the dashboard to make it easier to update in the future. You had some sort of optimisations implemented which were questionable as they required you to store in the dashboard data the code of all the students. I removed it, but now there will be a few more calls to `dataFn`.

lfaucon commented on ac/ac-autocode/src/Dashboard.jsx:
```diff
@@ -3,23 +3,152 @@
+    }));
+
+  const debugData = () => {
+    const o = Object.keys(props.users).reduce((acc, userId) => {
```
The viewer receives as props the list of users and instances. I think the dashboards should be based on instances. Logs also have the instanceId to be used in the merge function

Keylas committed:
Adjustments (instanceId, validateConfig)

Keylas commented:
The reason I was logging students' code was in case it would happen to become useful at some point (say the teacher would want to "probe" some code that have errors to get an idea of why his students are having trouble). This indeed led to weird test cases. The only sort of optimization it allowed was to not take into account multiple successive clicks on the buttons but even then it was not perfect.

Keylas commented:
- changed userId to insanceId because it does make more sense
remark: in preview, instances goes [0,1] while the log.instanceId are 1 and 2 so the graph is not correct.
- took Stian's #755 issue into account. It actually makes sense to have an activity without tests so functionalities that use it now verify that props.config.tests is not undefined (most problems came from the fact that they assumed it to be [] initially whereas it seems to be undefined). Added a validateConfig to check that a non empty solution is provided when there are tests.

Keylas committed:
typo
